File: frontend/src/ecs/world.ts
// Entity Component System for Dota 2 Toons

export type EntityID = string;

export interface Component {
  readonly componentId: string;
}

export interface Entity {
  readonly id: EntityID;
  components: Map<string, Component>;
  active: boolean;
}

export interface System {
  readonly name: string;
  update(dt: number, world: World): void;
}

export interface World {
  entities: Map<EntityID, Entity>;
  tick: number;
  deltaTime: number;
  
  createEntity(id?: EntityID): Entity;
  destroyEntity(id: EntityID): void;
  getEntity(id: EntityID): Entity | undefined;
  
  addComponent<T extends Component>(entityId: EntityID, component: T): void;
  removeComponent(entityId: EntityID, componentId: string): void;
  getComponent<T extends Component>(entityId: EntityID, componentId: string): T | undefined;
  hasComponent(entityId: EntityID, componentId: string): boolean;
  
  registerSystem(system: System): void;
  unregisterSystem(systemName: string): void;
  
  update(dt: number): void;
}

export function createWorld(): World {
  const entities = new Map<EntityID, Entity>();
  const systems: System[] = [];
  let currentTick = 0;
  let currentDeltaTime = 0;
  
  const entityCounter = { count: 0 };
  
  const generateId = (): EntityID => {
    entityCounter.count++;
    return `entity_${entityCounter.count}`;
  };
  
  return {
    entities,
    tick: currentTick,
    deltaTime: currentDeltaTime,
    
    createEntity(id?: EntityID): Entity {
      const entity: Entity = {
        id: id || generateId(),
        components: new Map(),
        active: true,
      };
      entities.set(entity.id, entity);
      return entity;
    },
    
    destroyEntity(id: EntityID): void {
      entities.delete(id);
    },
    
    getEntity(id: EntityID): Entity | undefined {
      return entities.get(id);
    },
    
    addComponent<T extends Component>(entityId: EntityID, component: T): void {
      const entity = entities.get(entityId);
      if (entity) {
        entity.components.set(component.componentId, component);
      }
    },
    
    removeComponent(entityId: EntityID, componentId: string): void {
      const entity = entities.get(entityId);
      if (entity) {
        entity.components.delete(componentId);
      }
    },
    
    getComponent<T extends Component>(entityId: EntityID, componentId: string): T | undefined {
      const entity = entities.get(entityId);
      if (entity) {
        return entity.components.get(componentId) as T | undefined;
      }
      return undefined;
    },
    
    hasComponent(entityId: EntityID, componentId: string): boolean {
      const entity = entities.get(entityId);
      if (entity) {
        return entity.components.has(componentId);
      }
      return false;
    },
    
    registerSystem(system: System): void {
      systems.push(system);
    },
    
    unregisterSystem(systemName: string): void {
      const index = systems.findIndex(s => s.name === systemName);
      if (index !== -1) {
        systems.splice(index, 1);
      }
    },
    
    update(dt: number): void {
      currentTick++;
      currentDeltaTime = dt;
      
      for (const system of systems) {
        system.update(dt, this);
      }
    },
  };
}

File: internal/game/world.go
// Package game implements the authoritative server-side ECS for Dota 2 Toons.
package game

import "sync"

// ---------------------------------------------------------------------------
// Entity
// ---------------------------------------------------------------------------

// EntityID is a unique string identifier for each entity.
type EntityID = string

// Entity holds a map of components keyed by their component type ID.
type Entity struct {
	ID         EntityID
	Components map[string]any
	Active     bool
}

// ---------------------------------------------------------------------------
// World
// ---------------------------------------------------------------------------

// System is executed every tick.
type System interface {
	Name() string
	Update(dt float64, w *World)
}

// World is the authoritative ECS world running on the server.
type World struct {
	mu       sync.RWMutex
	entities map[EntityID]*Entity
	systems  []System
	counter  int
	Tick     int
}

// NewWorld creates an empty world.
func NewWorld() *World {
	return &World{
		entities: make(map[EntityID]*Entity),
	}
}

// CreateEntity allocates a new entity and returns it.
func (w *World) CreateEntity() *Entity {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.counter++
	id := EntityID(w.generateID())
	e := &Entity{ID: id, Components: make(map[string]any), Active: true}
	w.entities[id] = e
	return e
}

func (w *World) generateID() string {
	// Simple incrementing ID — good enough for server authority
	return "entity_" + itoa(w.counter)
}

// DestroyEntity marks an entity as inactive.
func (w *World) DestroyEntity(id EntityID) {
	w.mu.Lock()
	defer w.mu.Unlock()
	if e, ok := w.entities[id]; ok {
		e.Active = false
	}
}

// GetEntity returns the entity or nil.
func (w *World) GetEntity(id EntityID) *Entity {
	w.mu.RLock()
	defer w.mu.RUnlock()
	return w.entities[id]
}

// Entities returns a snapshot of all active entities (safe to iterate).
func (w *World) Entities() []*Entity {
	w.mu.RLock()
	defer w.mu.RUnlock()
	out := make([]*Entity, 0, len(w.entities))
	for _, e := range w.entities {
		if e.Active {
			out = append(out, e)
		}
	}
	return out
}

// AddComponent stores a component on an entity.
func (w *World) AddComponent(id EntityID, componentID string, value any) {
	w.mu.Lock()
	defer w.mu.Unlock()
	if e, ok := w.entities[id]; ok {
		e.Components[componentID] = value
	}
}

// RemoveComponent deletes a component from an entity.
func (w *World) RemoveComponent(id EntityID, componentID string) {
	w.mu.Lock()
	defer w.mu.Unlock()
	if e, ok := w.entities[id]; ok {
		delete(e.Components, componentID)
	}
}

// HasComponent returns true if the entity has the given component.
func (w *World) HasComponent(id EntityID, componentID string) bool {
	w.mu.RLock()
	defer w.mu.RUnlock()
	if e, ok := w.entities[id]; ok {
		_, has := e.Components[componentID]
		return has
	}
	return false
}

// GetComponent returns the raw component value or nil.
func (w *World) GetComponent(id EntityID, componentID string) any {
	w.mu.RLock()
	defer w.mu.RUnlock()
	if e, ok := w.entities[id]; ok {
		return e.Components[componentID]
	}
	return nil
}

// RegisterSystem adds a system to the execution list.
func (w *World) RegisterSystem(s System) {
	w.systems = append(w.systems, s)
}

// Update runs all registered systems with the given delta time (seconds).
func (w *World) Update(dt float64) {
	w.Tick++
	for _, s := range w.systems {
		s.Update(dt, w)
	}
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	buf := [20]byte{}
	pos := len(buf)
	neg := n < 0
	if neg {
		n = -n
	}
	for n > 0 {
		pos--
		buf[pos] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		pos--
		buf[pos] = '-'
	}
	return string(buf[pos:])
}
